Make YouTube Embeds GDPR compliant

julieg18 opened the pull request:
* adds a message about youtube's policys when a user hovers over a youtube embed. 
* For our video component, I just added  a message below our overlay button. For our markdown, this is done by using a custom `gatsby-remark-embed` transformer and inserting a js script that handles taking off the message when the user has clicked. 

**To Do**

- [x] Review the js implementation I used for the md embeds
- [x] Decide if the design is good enough for a first iteration at least
- [x] Take off the embeds's message across the site if the user has already clicked on play for one of the embeds (we can do something like this with localStorage)

julieg18 commented on plugins/gatsby-theme-iterative-docs/src/components/Documentation/Markdown/Main/styles.module.css:
```diff
@@ -273,6 +273,53 @@
     margin-right: 10px;
   }
+
+  .yt-embed-wrapper {
```
I'm not sure if this design is good enough for a first iteration 🤔 

![image](https://user-images.githubusercontent.com/43496356/152210976-7eb931bd-c331-4677-ae06-4cc6bd1afc31.png)

shcheklein replied:
This looks great to me! This happens on hover, right?

julieg18 replied:
Yes!

https://user-images.githubusercontent.com/43496356/152249881-212056ca-1460-4ad7-97fd-3eae85226479.mov

julieg18 commented on plugins/gatsby-theme-iterative-docs/src/utils/front/setUpCustomYtEmbeds.ts:
```diff
@@ -0,0 +1,16 @@
+const setUpCustomYtEmbeds = () => {
```
This script takes all of our youtube embeds and attaches a click listener to our overlay. If the overlay is clicked, the video is played and the overlay disappears. It's definitely not very "reacty"... Does anybody have any ideas on how we could improve this?

Once we agree on a implementation (whether we deside to do this one or a different way) I want to have the code update local storage so that the message won't appear again if the user revisits the browser :)

rogermparent replied:
> Does anybody have any ideas on how we could improve this?

Bit of a shot in the dark, but is it possible to have the custom embedder return a custom tag like `<youtube>` and then attach a React component to our `rehype-react` custom components with this logic in it? That way we can use the `remark-embedder` style of automatic URL grabbing with React JS more consistent with the rest of the site.

julieg18 replied:
I could try that but wouldn't that only work for our docs pages? We'd probably want a solution that works the same for blog and doc 🤔

rogermparent replied:
Good point, I forgot about our weird divergent blog markdown.

yathomasi commented on config/gatsby-remark-embedder/custom-yt-embedder.js:
```diff
@@ -0,0 +1,59 @@
+const getHTML = url => {
+  const iframeSrc = getYouTubeIFrameSrc(url)
+
+  return `<div class="yt-embed-wrapper"><iframe width="100%" height="315" src="${iframeSrc}" frameBorder="0" allow="autoplay; encrypted-media; gyroscope; picture-in-picture" allowFullScreen></iframe><div className="yt-embed-wrapper__overlay"><span class="yt-embed-wrapper__tooltip">By clicking play, you agree to YouTube&apos;s <a href="https://policies.google.com/u/3/privacy?hl=en">Privacy Policy< ...
```
I am not sure but do we want to have the same height for all the youtube embeds?

julieg18 replied:
Yes, I think so. I used the same height that the youtube embeds were originally :)

yathomasi reviewed (commented):
Great going @julieg18!

rogermparent commented on config/gatsby-remark-embedder/custom-yt-embedder.js:
```diff
@@ -0,0 +1,59 @@
+  const { host, pathname, searchParams } = new URL(url)
+
+  return (
+    host === 'youtu.be' ||
+    (['youtube.com', 'www.youtube.com'].includes(host) &&
+      pathname.includes('/watch') &&
+      Boolean(searchParams.get('v')))
+  )
```
Why not add `'youtu.be'` to the `.includes` array?

julieg18 replied:
I copied this code over from `gatsby-remark-embed` so I'm not sure why the code is written like that. I'll update it and test if it works the same :)

yathomasi replied:
I also noticed and checked. I think it's due to the URL difference 
- `https://youtu.be/kLKBcPonMYw`
- `https://www.youtube.com/watch?v=kLKBcPonMYw`

julieg18 replied:
Updated the code and everything looks to be working!

rogermparent replied:
Good to know, I read the parentheses wrong, sorry for the false suggestion!

rogermparent commented on config/gatsby-remark-embedder/custom-yt-embedder.js:
```diff
@@ -0,0 +1,59 @@
+    return timeValue
+  }
+
+  const {
+    2: hours = '0',
+    4: minutes = '0',
+    6: seconds = '0'
+  } = timeValue.match(/((\d*)h)?((\d*)m)?((\d*)s)?/)
```
Minor nitpick, but you could use non-capturing groups (i.e. `/(?:(\d*)h)?(?:(\d*)m)?(?:(\d*)s)?/`) here to skip out on the otherwise useless group values and collect 1, 2, and 3 here.

julieg18 replied:
Done and worked the same as before!

rogermparent reviewed (commented):
I absolutely love the visual design on hover! Great stuff!

I think implementation could be improved a bit by my suggestion if possible, or if not at least @yathomasi's suggestion to use a hook for `setUpCustomYtEmbeds` (though I'm not sure how necessary it when events are added to elements in the body as opposed to `window` or `document`).

rogermparent commented:
```
$ tsc --noEmit --skipLibCheck && eslint --ext .json,.js,.ts,.tsx .
plugins/gatsby-theme-iterative-docs/src/utils/front/setUpCustomYtEmbeds.ts:9:57 - error TS2345: Argument of type 'EventTarget | null' is not assignable to parameter of type 'Node | null'.
  Type 'EventTarget' is missing the following properties from type 'Node': baseURI, childNodes, firstChild, isConnected, and 43 more.

9       if (event.target === tooltip || tooltip?.contains(event.target)) {
```

Maybe `if (event.target === tooltip || tooltip?.contains(event.target as Node)) {` would solve this, but maybe my raw API knowledge is rusty and something more needs to be done, as I really thought `.target` was just a `Node`.

julieg18 commented:
> Maybe if (event.target === tooltip || tooltip?.contains(event.target as Node)) { would solve this, but maybe my raw API knowledge is rusty and something more needs to be done, as I really thought .target was just a Node.

Thanks for that @rogermparent! I spend like half an hour trying to figure out how to clear that 🤣

julieg18 commented on src/components/Blog/Post/index.tsx:
```diff
@@ -38,6 +39,10 @@ const Post: React.FC<IBlogPostData> = ({
+  useEffect(() => {
+    setUpCustomYtEmbeds()
+  }, [])
+
```
Turned into a hook and added logic for removing event listeners.

julieg18 commented:
Updated the code based on everyone's comments and implemented `localStorage`! Once you click on a youtube embed (whether it's on the markdown or Video component), you won't see the message again on that browser.

yathomasi reviewed (commented):
The implementation looks great. 
Just a small thing I noticed on anchor links on tooltips on the different on the homepage, docs, and blogs, especially on different hover states.
[HomePage](https://dvc-org-update-yt-embed-71lby2.herokuapp.com/) - links are underlined
[Docs](https://dvc-org-update-yt-embed-71lby2.herokuapp.com/doc/start/data-and-model-versioning)- links are not underlined and will get underlined on hover
[Blogs](https://dvc-org-update-yt-embed-71lby2.herokuapp.com/blog/January-22-heartbeat) - links are not underlined and will get gray(opacity lower) on hover

I am not sure how much difference it makes as it is just a one-time view for users. But, if we want to be consistent I think it's good how we show on HomePage (the links are differentiated with underline)

rogermparent reviewed (approved):
Looks good to me as well, all that's left is what @yathomasi pointed out with the overlays' links.

julieg18 commented:
> Just a small thing I noticed on anchor links on tooltips on the different on the homepage, docs, and blogs, especially on different hover states.

Thanks for spotting that! Updated all the styles, keeping them all consistent and added a focus style.
